### PR TITLE
[WIP] Diffs for hydration errors

### PR DESCRIPTION
**Pull request by himself65**
## Summary

continue working on Upstream PR: https://github.com/facebook/react/pull/24250, Thanks the work by @gaearon 

## Progress

* [ ]  Errors
  * [x]  Extra server element
  * [x]  Extra server text
  * [x]  Extra client element
  * [x]  Extra client text
  * [ ]  Extra server attributes
  * [ ]  Prop mismatch
* [ ]  Formatting
  * [x]  Parent tag and attributes
  * [x]  Sibling tag/text and attributes
  * [x]  Child tag/text and attributes
  * [x]  Ellipsis to indicate more siblings before or after
* [ ]  Special cases
  * [ ]  Document/fragment parent
  * [x]  Long attributes
  * [x]  Too many attributes
  * [x]  Quotes in attributes
  * [ ]  Long text
  * [x]  Skip over non-element/text siblings
* [ ]  Other stuff
  * [ ]  Flow

### CODE REVIEW

**Comment by ilyaGurevich** [comment]
@himself65  just wanted to follow-up here, nytimes.com has recently migrated over to React 18 and are experiencing an influx of hydration mis-match errors that we hope to squash soon, however, the current stack trace is a bit hard to parse and we hope to be able to rely on this diffing PR as a reliable way to debug, do you have a sense of when this work might be completed! It would help a lot on our end.

**Comment by himself65** [comment]
> @himself65 just wanted to follow-up here, nytimes.com has recently migrated over to React 18 and are experiencing an influx of hydration mis-match errors that we hope to squash soon, however, the current stack trace is a bit hard to parse and we hope to be able to rely on this diffing PR as a reliable way to debug, do you have a sense of when this work might be completed? It would help a lot on our end.

Sorry for late update on this PR. Currently im working on llamaindex and other react downstream libraries, I think I can continue on this PR during this month. Logic is simple but Mostly Im stucking on debugging flow and react source code.

Sorry for low contribution on this but Im trying

**Comment by himself65** [comment]
@ilyaGurevich I will recommend using this for hydration diff for a fast fix for now.

https://github.com/BuilderIO/hydration-overlay

**Comment by sebmarkbage** [comment]
Thanks for picking this up! We're going with a different approach with a more invasive refactor in #28512. I'll close this one out.
